File: leetcode/cpp/number-of-provinces.cpp
/* -------------------------------------------------------------------------- */
/*              https://leetcode.com/problems/number-of-provinces             */
/* -------------------------------------------------------------------------- */
#include <stack>
#include <vector>
#include <iostream>

using namespace std;

int findCircleNum(vector<vector<int>>& isConnected) {
    int num_nodes = isConnected.size();
    auto visited = vector<bool>(num_nodes, false);

    int num_components = 0;
    for (int node = 0; node < num_nodes; node++) {
        if (visited[node]) {
            continue;
        }

        num_components++;

        stack<int> st;
        st.push(node);

        while (!st.empty()) {
            int top = st.top();
            st.pop();
            visited[top] = true;

            for (int nbr = 0; nbr < num_nodes; nbr++) {
                if (nbr != top && !visited[nbr] && isConnected[top][nbr] == 1) {
                    st.push(nbr);
                }
            }
        }
    }
    return num_components;
}

File: leetcode/rust/src/number_of_provinces.rs
/* ------------ https://leetcode.com/problems/number-of-provinces ----------- */

pub fn find_circle_num(is_connected: Vec<Vec<i32>>) -> i32 {
    let num_nodes = is_connected.len();
    let mut visited = vec![false; num_nodes];

    let mut num_components = 0;
    for node in 0..num_nodes {
        if visited[node] {
            continue;
        }

        num_components += 1;
        let mut stack = vec![];
        stack.push(node);
        while let Some(top) = stack.pop() {
            visited[top] = true;
            for nbr in 0..num_nodes {
                if nbr != node && !visited[nbr] && is_connected[top][nbr] == 1 {
                    stack.push(nbr)
                }
            }
        }
    }

    num_components
}
